[CMS-232] Add general purpose helper methods to Site and Upstream and refactor clone:local command

jabxjab opened the pull request.

Commit:
Add framework-related getters to Site class

Commit:
Add isDrops8Upstream() to SiteUpstream

Commit:
Fix Site::getUpstream() declaration

jabxjab commented on src/Models/Site.php:
```diff
@@ -234,6 +234,46 @@ public function getName()
+     *
+     * @return bool
+     */
+    public function isDrupal8Framework(): bool
```
Used here https://github.com/pantheon-systems/terminus-conversion-tools-plugin/blob/main/src/Commands/ConvertToComposerSiteCommand.php#L30

jabxjab commented on src/Models/SiteUpstream.php:
```diff
@@ -55,12 +56,22 @@ protected function parseAttributes($data)
+     *
+     * @return bool
+     */
+    public function isDrops8Upstream(): bool
```
Used here https://github.com/pantheon-systems/terminus-conversion-tools-plugin/blob/main/src/Commands/ConvertToComposerSiteCommand.php#L37

greg-1-anderson replied:
I don't think that Terminus should be in the business of knowing upstream machine names. Move this over to the conversion tools plugin. Rather than having a class with a lot of is-this-or-that-upstream, I think we could just define constants (e.g. public class constants) to compare against. I care 1 out of 10, though, so you can keep the implementation as "is" methods if you really feel that is better.

jabxjab replied:
removed

greg-1-anderson commented on src/Models/Site.php:
```diff
@@ -234,6 +234,46 @@ public function getName()
+     */
+    public function isWordpressFramework(): bool
+    {
+        return 'wordpress' === $this->getFramework();
```
This should probably test to see if the framework begins with `wordpress`, to account for `wordpress-network`. Consider `isWordPressNetworkFramework()` and `isDrupalFramework()`.

jabxjab replied:
fixed

greg-1-anderson commented on src/Models/Site.php:
```diff
@@ -234,6 +234,46 @@ public function getName()
+    }
+
+    /**
+     * Returns TRUE if the site framework is Drupal 7.
```
Note in comment that this value is for Drupal 6 and Drupal 7.

greg-1-anderson commented on src/Models/Site.php:
```diff
@@ -234,6 +234,46 @@ public function getName()
+    }
+
+    /**
+     * Returns TRUE if the site framework is Drupal 8.
```
Note in comment that this value is for Drupal 8 and Drupal 9.

greg-1-anderson reviewed (changes requested):
These methods are very useful, but I am ambivalent about having so many framework-related helper methods in the Site model. What do you think about the idea of putting only `getFramework()` here, and then create a framework utility class that has all of the different comparators in it?

Commit:
Add SiteFramework model

jabxjab commented:
> These methods are very useful, but I am ambivalent about having so many framework-related helper methods in the Site model. What do you think about the idea of putting only `getFramework()` here, and then create a framework utility class that has all of the different comparators in it?

Agree, sounds reasonable. Added `SiteFramework` model for that

Commit:
Refactor CommandExecutorTrait

Commit:
Change Site::getLocalCopyDir() to accept optional $siteDirName

Commit:
Refactor local:clone command

* update to better error handling
* add "site_dir" and "override" options
* fix code style issues
* remove redundant code

jabxjab commented:
@greg-1-anderson @namespacebrian I've also added a refactoring of `clone:local` command:
* added `site_dir` and `override` options (will be utilized by `terminus-conversion-tools-plugin`)
* fixed code style issues
* removed redundant code
* improved error handling 
(https://github.com/pantheon-systems/terminus-conversion-tools-plugin/pull/1)

Commit:
Fix inconsistency in site loading by name vs. uuid

* fixed false positive site load by uuid in the scenario where the site exists but not available to the authenticated user
* improved UUID value verification algorithm (must comply strictly with v4 UUID standard, added "start" and "end" anchors)
* removed redundant code parts
* fixed code styles

jabxjab commented:
Added a fix to inconsistent behavior of site object loading by name vs. uuid (reproducible as false positive by executing `site:info` against an existing but not available to the authenticated terminus user site)

Commit:
Create a generic cloneGitRepository() in LocalMachineHelper and use it in local:clone command

Commit:
Change CommandExecutorTrait::execute() argument names

greg-1-anderson commented on src/Collections/Sites.php:
```diff
@@ -261,33 +287,40 @@ protected function getUuidByName(string $name): string
+     * @throws \GuzzleHttp\Exception\GuzzleException
      */
-    protected function findUUIDByNameOrUUID($id): ?string
+    protected function getUuidBySiteId(string $site_id): ?string
```
The new name isn't as clear as the old one. The docblock still stipulates "name or uuid", but this is less clear when reading the code. It's protected & no one should be extending sites, but the old name seemed fine. I'm indifferent between get and find (and camelcase conventions), but maybe keep 'NameOrUuid'?

jabxjab replied:
There are many places where we have "name or uuid" for the `$site_id` argument without that extra suffix ("get by something"). So decided to rename it into a generic `getUuid()`.

greg-1-anderson commented on src/Collections/Sites.php:
```diff
@@ -261,33 +287,40 @@ protected function getUuidByName(string $name): string
+    protected function isValidUuidV4(string $uuid): bool
     {
-        return preg_match('/[a-f0-9]{8}-([a-f0-9]{4}-){3}[a-f0-9]{12}/', strtolower($id));
+        return preg_match('/^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/', strtolower($uuid));
```
Do you know whether our platform always generates UUIDs that are strictly v4 conforming? It might be good to find the code that generates them to be sure. This is probably in ygg. Alternately, you probably wouldn't need to test too many actual UUIDs, if they are not conforming, it wouldn't take too long to find one that didn't validate.

jabxjab replied:
Reverting back the existing logic to prevent potential incompatibility

greg-1-anderson commented on src/Models/SiteFramework.php:
```diff
@@ -0,0 +1,43 @@
+ *
+ * @package Pantheon\Terminus\Models
+ */
+class SiteFramework extends TerminusModel
```
This should not be a TerminusModel. Just make it a plain old PHP class. Using this class will be a two-step process:

1. Get the framework value from the Site object
2. Instantiate a new SiteFramework utility object via `new SiteFramework($framework)`

greg-1-anderson replied:
After making this change, also move this class to another namespace, e.g. maybe Utils.

jabxjab replied:
fixed

greg-1-anderson commented on src/Models/Site.php:
```diff
@@ -234,6 +234,23 @@ public function getName()
+            $nickname = \uniqid(__FUNCTION__ . "-");
+            $this->getContainer()->add($nickname, SiteFramework::class)
+                ->addArgument((object) ['framework' => $this->get('framework')]);
+            $this->framework = $this->getContainer()->get($nickname);
```
See comment below; SiteFramework should just be instantiated via `new`.

greg-1-anderson reviewed (changes requested):
Looks pretty good, just a couple of minor style comments.

Commit:
Rename getUuidBySiteId() into a more generic getUuid()

Commit:
Revert isValidUuid() logic back

Commit:
Move SiteFramework into Pantheon\Terminus\Helpers\Utility

Commit:
Remove isDrops8Upstream()

greg-1-anderson reviewed (approved):
LGTM
